### DISCUSSION TITLE
Support for Binary + ULID IDs in Rodauth

### DISCUSSION THREAD

**Discussion by ukmshi** [jeremyevans/rodauth#392] [General]
Dear Rodauth Team,

First and foremost, I'd like to express my gratitude for developing Rodauth. Its ease of use and robustness have been greatly appreciated in my projects.

## Background

Currently, I am developing a Rails application utilizing MySQL with Binary + ULID for generating unique identifiers. I am keen on implementing this ID format in the tables required by Rodauth as well, to maintain consistency throughout the application.

## Challenge
However, I've encountered a challenge with Rodauth when adopting Binary + ULID. The direct use of Binary values for IDs leads to errors during transformations, such as when generating JWTs.

## Proposal

In Rails, it's possible to declare a custom type via `ActiveRecord::Type::Binary`, which allows for altering the processing between the database and `ActiveRecord` during reads and writes, as demonstrated below:

```ruby
class ULIDType < ActiveRecord::Type::Binary
  # Convert the value before writing to the database (no conversion needed here, so it returns the value as is)
  def cast(value)
    value
  end
  
  # Convert the database read value into a Ruby object
  def deserialize(value)
    raw_value = super(value) # Retrieve the binary data from the database
    return if raw_value.nil?

    Utils::Ulid.binary_to_ulid_string(raw_value.to_s)
  end

  # Convert the Ruby object into a format for database storage
  def serialize(value)
    return if value.nil?

    raw_value = Utils::Ulid.string_to_ulid_byte(value)
    super(raw_value)
  end
end

ActiveRecord::Type.register(:ulid, ULIDType)
```

Does Rodauth have a mechanism for customizing ID reads and writes, similar to `ActiveRecord::Type`?

**Comment by jeremyevans** [comment]
Rodauth uses Sequel for database access, and Sequel doesn't have a mechanism similar to `ActiveRecord::Type`.  Type conversion in Sequel is customized per adapter.  The issue with MySQL is that the drivers that Sequel uses (mysql2, trilogy), do not support customizing the type conversion, so the type that Sequel would usually use for binary fields (Sequel::SQL::Blob), isn't used for those adapters.

So currently, I don't think there is an easy way to support what you want.  Rodauth is designed to work both with integer and UUID keys.  If you were using PostgreSQL, there is an actual uuid type that stores in binary but uses text for input/output, so that would work without changes to Rodauth.

**Reply by ukmshi** [reply]
@jeremyevans 
Thank you for your helpful response. I have decided to go with using integer primary keys.
